src/iconv.jl

```diff
@@ -1,20 +1,76 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
 module iconv
-import Base: close, eof, flush, read, readall, write
+import Base: close, eof, flush, read, readall, write, show
 import Base.Libc: errno, strerror, E2BIG, EINVAL, EILSEQ
 export StringEncoder, StringDecoder, encode, decode
+export StringEncodingError, GeneralError
+export InvalidEncodingError, InvalidSequenceError, IncompleteSequenceError
+
+const UNKNOWN_ERR      = "<<1>> (<<2>>)"
+const OUTPUT_BUFFER    = "ran out of space in the output buffer"
+const UNSUPPORTED_CONV =
+    "conversion from <<1>> to <<2>> not supported by iconv implementation, check that specified encodings are correct"
+const INVALID_SEQUENCE    = "byte sequence 0x<<1>> is invalid in source encoding or cannot be represented in target encoding"
+const INCOMPLETE_SEQUENCE = "incomplete byte sequence at end of input"
+const INVALID_BYTES       = "invalid byte sequence in input"
+
+abstract StringEncodingError
+
+# General purpose error handling type, which supports variable number of arguments,
+# and can be extended to work with different locales
+
+type GeneralError <: Exception
+    errcls::UTF8String
+    errmsg::UTF8String
+    args
+    GeneralError(class::AbstractString, msg::AbstractString, args...) =
+        new(UTF8String(class), UTF8String(msg), args)
+end
+
+function show(io::IO, exc::GeneralError)
+    contains(exc.errmsg, "<<") || return print(io, exc.errcls, ": ", exc.errmsg)
+    str = exc.errmsg
+    for i = 1:length(exc.args)
+        str = replace(str, "<<$i>>", exc.args[i])
+    end
+    print(io, exc.errcls, ": ", str)
+end
+
+const modulename = "iconv"
+
+# specified encodings are not supported
+type InvalidEncodingError <: StringEncodingError
+    err::GeneralError
+    InvalidEncodingError(args...) = new(GeneralError(modulename, UNSUPPORTED_CONV, args...))
+end
+
+# Encountered invalid byte sequence
+type InvalidSequenceError <: StringEncodingError
+    err::GeneralError
+    InvalidSequenceError(msg, args...) = new(GeneralError(modulename, msg, args...))
+end
+
+# Input ended with incomplete byte sequence
+type IncompleteSequenceError <: StringEncodingError
+    err::GeneralError
+    IncompleteSequenceError() = new(GeneralError(modulename, INCOMPLETE_SEQUENCE))
+end
+
+show{T<:StringEncodingError}(io::IO, exc::T) = show(io, exc.err)
+
+generalerror(class) =
```

Comment:
Better call this e.g. `iconv_error` and have it create an object of type `IConvError <: StringEncodingError`. You could have one field giving the name of the function, and another one the error code.